### PR TITLE
Device: Don't remove an existing target directory when unmounting a `disk` device if the original dir hasn't been created by LXD 

### LINKED ISSUE #12648: Removing an LXD disk device deletes an empty directory it was bind mounted over
# Required information

 * Distribution: ubuntu
 * Distribution version: 23.04
 * The output of "lxc info" or if that fails:
   * Kernel version:
   * LXC version:
   * LXD version: 5.19-8635f82
   * Storage backend in use: dir

# Issue description

LXD behaves inconsistently when it deals with bind-mounts on empty instance's directories. If an instance is running and one removes a disk device that was bind mounted over an existing empty directory, the underlying directory gets removed as well. Conversely, when the instance is stopped, removing the disk device would not lead to removing the instance's empty directory.

As per the [source code](https://github.com/canonical/lxd/blob/e90ae16cc7c573d5a6ad37783e888190afda9ffd/lxd/instance/drivers/driver_lxc.go#L1744), that seems to be a designed behaviour as LXD creates the target (path) directories if those did not exist on the host. However, it leads to removing directories that were not created by LXD.

# Steps to reproduce

Note: /opt is an existing and empty directory.

```
lxc launch ubuntu:22.04 craft-0758614d
lxc start craft-0758614d
lxc config device add craft-0758614d test disk source=/home/dmitry/test path=/opt
lxc config device remove craft-0758614d test
lxc exec craft-0758614d -- ls /opt
ls: cannot access '/opt': No such file or directory
```

### LINKED ISSUE #12716: Removing a `disk` device with an existing target path on a VM leads to unreachable target directory
# Required information
Distribution: ubuntu
Distribution version: 22.04
The output of "lxc info" or if that fails:
Kernel version:
LXC version:
LXD version: 5.19-8635f82
Storage backend in use: zfs

# Issue description
Similarly to https://github.com/canonical/lxd/issues/12648, but for a VM instance this time, we observe an issue while removing a disk device that was bind mounted over an existing empty directory. This time, the underlying directory doesn't get removed but cannot but interacted with :

```bash
mkdir /tmp/empty-dir
lxc launch ubuntu:jammy v1 --vm
lxc config device add v1 test disk source=/tmp/empty-dir path=/opt
lxc config device remove v1 test
lxc exec v1 -- ls /opt
ls: cannot access '/opt': Transport endpoint is not connected
```

I would expect that, for an already exiting directory, the target directory would be unmounted but still present, whereas a target directory created through LXD during the mount woulld be unmounted and deleted when unmounted. 

We should also check that, within the VM, the target path we're attempting to unmount has no overmounts as calling `umount` on the target directory will first unmount the last mount but not all of them.

### PR DESCRIPTION
**Pull request by gabrielmougard**
closes #12648 
closes #12716
____

# Overall description

When a disk device is removed while relying on a host directory and mapped to a target within a container or a VM,
we detect if the target directory has been created by LXD or not in order to not delete the content of a target
directory during the unmount. Additionaly, with the VM case, we cleanly unmount the target path inside the VM.  

# Container case

Example:

- Let say we mounted a custom host empty directory (`test`) on the existing `/opt` directory of the container,
when unmounted (`lxc device remove ...`), the target `/opt` won't be removed, because we marked it as NOT being created
by LXD at mount time.

- If, however, we created a custom empty target directory at mount time:
`lxc config device add u1 test disk source=/home/user/test path=/new_dir`,
the directory `new_dir` will be created on the target instance and if we decide to unmount `test`,
the target `/new_dir` will be removed because is has been created by LXD

# Particular case for VMs

In addition to that, this fix also cleanly unmount the target directory inside the VM through an new LXD-agent API call. Before that, here is what would have happened:

```bash
mkdir /tmp/empty-dir
lxc launch ubuntu:jammy v1 --vm
lxc config device add v1 empty-dir disk source=/tmp/empty-dir path=/opt
lxc config device remove v1 empty-dir

lxc shell v1 -- stat /opt
stat: cannot statx '/opt': Transport endpoint is not connected
``` 

This happens because the mounted device and its associated char device were removed using QEMU's QMP without unmounting the target.

Now, we inspect for the mounts and the over-mounts if any on the VM, and unmount them in the right order. 

## Benchmark

Each time, there are 10 runs on each benchmark case (10 starts, 10 stops)

`master` branch (CONTAINER):

Startup time with:                   Stop time (`--force`) with:

* 0 disk:    `598.2 ms ±  14.3 ms`   * 0 disk:    `883.5 ms ±  37.1 ms`
* 5 disk:    `595.0 ms ±   8.2 ms`   * 5 disk:    `892.6 ms ±  23.3 ms`
* 10 disks:  `601.0 ms ±   8.5 ms`   * 10 disks:  `900.1 ms ±  22.0 ms`
* 20 disks:  `606.9 ms ±  12.7 ms`   * 20 disks:  `886.8 ms ±  23.4 ms`
* 100 disks: `634.3 ms ±  10.1 ms`   * 100 disks: `934.1 ms ±  16.4 ms`

Our fix branch (CONTAINER):

Startup time with:                   Stop time (`--force`) with:

* 0 disk:    `632.8 ms ±  21.3 ms`   * 0 disk:    `870.3 ms ±  25.6 ms`
* 5 disk:    `624.4 ms ±  15.4 ms`   * 5 disk:    `879.9 ms ±  23.6 ms`
* 10 disks:  `626.6 ms ±  17.2 ms`   * 10 disks:  `877.1 ms ±  19.7 ms`
* 20 disks:  `641.0 ms ±  23.3 ms`   * 20 disks:  `932.0 ms ± 148.6 ms`
* 100 disks: `660.0 ms ±  18.3 ms`   * 100 disks: `942.7 ms ±  16.9 ms`

### CODE REVIEW

**Comment by gabrielmougard** [comment]
Working on the integration tests now.

**Comment by gabrielmougard** [comment]
Tests should be ready

File: shared/instance.go [hunk 1]
```diff
@@ -1206,6 +1206,15 @@ func ConfigKeyChecker(key string, instanceType instancetype.Type) (func(value st
 			return validate.IsAny, nil
 		}
 
+		// lxdmeta:generate(entity=instance, group=volatile, key=volatile.<name>.<target_path>.dir_created_by_lxd)
+		//
+		// ---
+		//  type: bool
+		//  shortdesc: if the disk device mounted from a host directory, created a new directory within the instance
```
Comment (roosterfish) [inline shared/instance.go:1213]: `shortdesc: If the ...`

File: shared/instance.go [hunk 2]
```diff
@@ -1206,6 +1206,15 @@ func ConfigKeyChecker(key string, instanceType instancetype.Type) (func(value st
...
+		if strings.HasSuffix(key, ".dir_created_by_lxd") {
```
Comment (roosterfish) [inline shared/instance.go:1214]: I feel `dir_created_by_lxd` looks to be quite long compared to the other volatile keys. Maybe we could just do `dir_created`?

File: test/suites/container_devices_disk.sh [hunk 3]
```diff
@@ -86,6 +86,18 @@ test_container_devices_disk_shift() {
...
+  lxc config device add foo empty-dir disk source="${TEST_DIR}/empty-dir" path=/opt
+  lxc config device remove foo empty-dir
+  lxc exec foo -- stat /opt
+
+  lxc exec foo -- touch /tmp/empty-dir
+  lxc config device add foo empty-dir disk source="${TEST_DIR}/empty-dir" path=/tmp/empty-dir
+  lxc config device remove foo empty-dir
+  lxc exec foo -- stat /tmp/empty-dir
+  rm -rf "${TEST_DIR}/empty-dir"
+
```
Comment (roosterfish) [inline test/suites/container_devices_disk.sh:100]: Please also add a third test here checking if a non existing directory gets removed after the disk device gets removed from the container (`.dir_created_by_lxd=true`). Then every case should be covered.

File: doc/config_options.txt [hunk 4]
```diff
@@ -778,6 +778,12 @@ Specify either a cron expression (`<minute> <hour> <dom> <month> <dow>`), a comm
 
 <!-- config group instance-snapshots end -->
 <!-- config group instance-volatile start -->
+```{config:option} volatile.<name>.<target_path>.dir_created_by_lxd instance-volatile
```
Comment (roosterfish) [inline doc/config_options.txt:781]: Do we required `name` and `target_path`? Or could the device name alone be sufficient? If you name the device `mnt` and mount it to `mnt` inside the container you get `volatile.mnt.mnt.dir_created_by_lxd` which might be redundant.

**Review by roosterfish — CHANGES_REQUESTED** [review]

File: lxd/instance/drivers/driver_lxc.go [hunk 5]
```diff
@@ -1713,6 +1713,31 @@ func (d *lxc) deviceHandleMounts(mounts []deviceConfig.MountEntryItem) error {
...
+				if err != nil {
+					return err
+				}
+
+				defer func() { _ = files.Close() }()
+
+				_, err = files.Lstat(mount.TargetPath)
+				if err == nil {
+					// The target path already exist.
+					err = d.deviceVolatileSetFunc(mount.DevName)(map[string]string{fmt.Sprintf("%s.dir_created_by_lxd", mount.TargetPath): "false"})
```
Comment (tomponline) [inline lxd/instance/drivers/driver_lxc.go:1729]: Lets use `%s.last_state.created` so we align with the NIC volatile keys that indicate something similar

Comment (tomponline) [reply lxd/instance/drivers/driver_lxc.go:1729]: Lets have a constant for this too so we can see where its referred to.

File: lxd/instance/drivers/driver_lxc.go [hunk 6]
```diff
@@ -1713,6 +1713,31 @@ func (d *lxc) deviceHandleMounts(mounts []deviceConfig.MountEntryItem) error {
...
+					if err != nil {
+						return fmt.Errorf("Error updating volatile for the device: %s", err)
```
Comment (tomponline) [inline lxd/instance/drivers/driver_lxc.go:1731]: wrap errors with %w please

File: shared/instance.go (hunk 1 above)
Comment (ru-fu) [inline shared/instance.go:1213]: Please split that up into a short description and a longer description. If you look at the output, the short description takes two lines (which is not "short" ;) ).
Also, I don't really understand what the option does from this description. Try to go for just the essence (maybe "How the host directory was created" or something?) for the short description, and then clearly describe what it means in the longer description.

File: test/suites/container_devices_disk.sh (hunk 3 above)
Comment (gabrielmougard) [reply test/suites/container_devices_disk.sh:100]: Do you mean a non-existing directory on the host side or on the target-side (inside the container) ?

Comment (roosterfish) [reply test/suites/container_devices_disk.sh:100]: Inside of the container. I see the first test checks that an existing dir in the container doesn't get removed. The second test checks if mounting over an existing file with the same name doesn't remove the file. And a third test could check if a directory gets removed if it was created by LXD explicitly for the mount.

File: doc/config_options.txt (hunk 4 above)
Comment (gabrielmougard) [reply doc/config_options.txt:781]: Yes, I can try to chhange the logic a bit to remove `target_path` from the key.

**Comment by gabrielmougard** [comment]
Also, @tomponline, the `deviceVolatileSetFunc` function, which calls the `VolatileSet`, updates the DB. I think this is good to persist this information though. Removing an existing target dir like `/opt` for example (see the original post of the issue author) through an umount seems quite bad.

Or, we can introduce a new device option to say "not to track that kind of changes" so that I can set `d.volatileSetPersistDisable = true` before my volatile set call and then restore it to `false` after. This will surely be faster (no DB call) but the user accept a potential data loss on the target instance if he mounted on an existing target dir.

File: doc/config_options.txt [hunk 7]
```diff
@@ -808,6 +808,14 @@ The network device MAC address is used when no `hwaddr` property is set on the d
 Possible values are `true` or `false`.
 ```
 
+```{config:option} volatile.<name>.last_state.created_path instance-volatile
+:shortdesc: "Path of the newly created target directory mounted from a `disk` device"
```
Comment (ru-fu) [inline doc/config_options.txt:812]: Still too long :(
```suggestion
:shortdesc: "Path of the directory created from mounting a `disk` device"
```

File: doc/config_options.txt [hunk 8]
```diff
@@ -808,6 +808,14 @@ The network device MAC address is used when no `hwaddr` property is set on the d
...
+:type: "string"
+Path of the created target directory that might have been mounted from a host directory by a `disk` device.
+For example, if the target directory is `/opt` and that it already exists in the instance, then this key will not be set.
+However, if the target directory is `/opt/foo` and that it doesn't exist in the instance, then this key will be set to `/opt/foo`.
```
Comment (ru-fu) [inline doc/config_options.txt:816]:
```suggestion
If mounting a host directory through a `disk` device creates the target directory, this option contains the path of this newly created target directory.
For example, if the target directory is `/opt` and `/opt` already exists in the instance, this key is not set.
However, if the target directory is `/opt/foo` and `/opt/foo` doesn't exist in the instance, this key is set to `/opt/foo`.
```

File: doc/api-extensions.md [hunk 9]
```diff
@@ -2329,3 +2329,7 @@ This introduces the configuration keys `cephfs.create_missing`, `cephfs.osd_pg_n
 
 This API extension provides the ability to use flags `--profile`, `--no-profile`, `--device`, and `--config`
 when moving an instance between projects and/or storage pools.
+
+## `existing_target_dir_mounted_by_disk_device`
+
+This API extension provides the ability to check if a target directory was created within the instance filesystem at mount time. If a host directory is mounted to an existing container directory (e.g., `/opt`), the target directory won't be removed upon unmounting. However, if LXD creates a new target directory during the mount, like `/new_dir`, it will be deleted when the device is unmounted.
```
Comment (ru-fu) [inline doc/api-extensions.md:2335]:
```suggestion
This API extension provides the ability to check if a target directory was created within the instance file system at mount time. If a host directory is mounted to an existing container directory (e.g., `/opt`), the target directory won't be removed upon unmounting. However, if LXD creates a target directory during the mount, like `/new_dir`, it will be deleted when the device is unmounted.
```

**Review by ru-fu — DISMISSED** [review]
Docs look good now. :)

**Comment by gabrielmougard** [comment]
@tomponline I implemented the logic for the VM as well within the lxd-agent. I tested it on my side without an issue.

**Comment by tomponline** [comment]
Thanks @gabrielmougard please can you add a PR description with what has changed so the reviewer can compare the code to the intended design. Cheers

**Comment by tomponline** [comment]
Please can you rebase @gabrielmougard

File: shared/version/api.go [hunk 10]
```diff
@@ -395,6 +395,7 @@ var APIExtensions = []string{
 	"init_preseed_storage_volumes",
 	"metrics_instances_count",
 	"server_instance_type_info",
+	"existing_target_dir_mounted_by_disk_device",
```
Comment (tomponline) [inline shared/version/api.go:398]: looking at other disk related api extensions they start with `disk_` so we should keep to the same convention.

Suggest `disk_state_created`

**Review by tomponline — COMMENTED** [review]
@gabrielmougard when the `disk` device is hot-plugged its mount inside the guest is triggered by an event in `eventsProcess()`. Is there a reason you didn't feel that we could unmount the disk in the same fashion and required the API endpoint instead?

File: lxd/instance/instancetype/instance.go [hunk 11]
```diff
@@ -1191,6 +1191,17 @@ func ConfigKeyChecker(key string, instanceType Type) (func(value string) error,
...
 		}
 
+		// lxdmeta:generate(entity=instance, group=volatile, key=volatile.<name>.last_state.created_path)
+		// If mounting a host directory through a `disk` device creates the target directory, this option contains the path of this newly created target directory.
+		// For example, if the target directory is `/opt` and `/opt` already exists in the instance, this key is not set.
+		// However, if the target directory is `/opt/foo` and `/opt/foo` doesn't exist in the instance, this key is set to `/opt/foo`.
+		// ---
+		//  type: string
+		//  shortdesc: Path of the directory created from mounting a `disk` device
+		if strings.HasSuffix(key, ".last_state.created_path") {
```
[the remark this answers is not in the document]
Comment (gabrielmougard) [reply lxd/instance/instancetype/instance.go:1201]: Oh sorry for that. I'm changing that

**Comment by gabrielmougard** [comment]
@tomponline that's interesting! I just didn't know about this `eventsProcess()` logic tbh. You're right, if we could handle the unmount logic as part of an event hook, that'd be better and it'll avoid adding new API endpoint. I'll see if I can move the logic.

**Comment by gabrielmougard** [comment]
@tomponline I moved the unmount logic bit into the `eventsProcess()` function as you required. I reproduced the integration tests on my side with a VM, so issue so far.

File: doc/api-extensions.md [hunk 12]
```diff
@@ -2348,3 +2348,7 @@ This API extension provides the ability to use flags `--device` when importing a
 ## `instances_uefi_vars`
 
 This API extension indicates that the `/1.0/instances/{name}/uefi-vars` endpoint is supported on the server. This endpoint allows to get the full list of UEFI variables (HTTP method GET) or replace the entire set of UEFI variables (HTTP method PUT).
+
+## `existing_target_dir_mounted_by_disk_device`
```
Comment (tomponline) [inline doc/api-extensions.md:2352]: commit message and docs mismatch

File: lxd/instance/drivers/driver_lxc.go [hunk 13]
```diff
@@ -1713,10 +1721,18 @@ func (d *lxc) deviceHandleMounts(mounts []deviceConfig.MountEntryItem) error {
...
+					return fmt.Errorf("Error updating volatile for the device: %w", err)
+				}
+			}
+
 			// Mount it into the container.
-			err := d.insertMount(mount.DevPath, mount.TargetPath, mount.FSType, flags, idmapType)
+			err = d.insertMount(mount.DevPath, mount.TargetPath, mount.FSType, flags, idmapType)
 			if err != nil {
-				return fmt.Errorf("Failed to add mount for device inside container: %s", err)
+				return fmt.Errorf("Failed to add mount for device inside container: %w", err)
```
Comment (tomponline) [inline lxd/instance/drivers/driver_lxc.go:1735]: This change should be a separate commit its not related to the commit message

File: lxd/instance/drivers/driver_lxc.go [hunk 14]
```diff
@@ -1732,17 +1748,36 @@ func (d *lxc) deviceHandleMounts(mounts []deviceConfig.MountEntryItem) error {
...
 			_, err = files.Lstat(relativeTargetPath)
 			if err == nil {
-				err := d.removeMount(mount.TargetPath)
-				if err != nil {
-					return fmt.Errorf("Error unmounting the device path inside container: %s", err)
+				removeTargetFiles := false
+
+				// Check if the target path hasn't been created by LXD
+				mountConf := d.deviceVolatileGetFunc(mount.DevName)()
+				for k, v := range mountConf {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_lxc.go:1755]: why iterating a map if you're looking up by key and comparing the value?

File: lxd/instance/drivers/driver_lxc.go [hunk 15]
```diff
@@ -1732,17 +1748,36 @@ func (d *lxc) deviceHandleMounts(mounts []deviceConfig.MountEntryItem) error {
...
+				if removeTargetFiles {
+					err = files.Remove(relativeTargetPath)
+					if err != nil {
+						// Only warn here and don't fail as removing a directory
+						// mount may fail if there was already files inside
+						// directory before it was mouted over preventing delete.
+						d.logger.Warn("Could not remove the device path inside container", logger.Ctx{"err": err})
+					}
+				} else {
+					// Remove option from the mount.
```
Comment (tomponline) [inline lxd/instance/drivers/driver_lxc.go:1776]: from device not mount

File: lxd-agent/events.go [hunk 16]
```diff
@@ -141,15 +140,78 @@ func eventsProcess(event api.Event) {
...
+		mntSource := fmt.Sprintf("lxd_%s", e.Name)
 
-	_ = os.MkdirAll(e.Config["path"], 0755)
-	_, err = shared.RunCommand("mount", "-t", "virtiofs", mntSource, e.Config["path"])
-	if err != nil {
-		logger.Infof("Failed to mount hotplug %q (Type: %q) to %q", mntSource, "virtiofs", e.Config["path"])
+		_ = os.MkdirAll(e.Config["path"], 0755)
+		_, err = shared.RunCommand("mount", "-t", "virtiofs", mntSource, e.Config["path"])
+		if err != nil {
+			logger.Infof("Failed to mount hotplug %q (Type: %q) to %q", mntSource, "virtiofs", e.Config["path"])
```
Comment (tomponline) [inline lxd-agent/events.go:150]: Shall we make this a logger.Warnf?

File: lxd-agent/events.go [hunk 17]
```diff
@@ -141,15 +140,78 @@ func eventsProcess(event api.Event) {
...
+			return
+		}
+
+		logger.Infof("Mounted hotplug %q (Type: %q) to %q", mntSource, "virtiofs", e.Config["path"])
+	} else if e.Action == "removed" {
+		targetPath := e.Config["path"]
+
+		if !filepath.IsAbs(targetPath) {
+			logger.Infof("the device path must be absolute: %q", e.Config["path"])
```
Comment (tomponline) [inline lxd-agent/events.go:159]: Shall we make this a logger.Warnf?

Also why lower case `t` at start of sentence?

File: lxd-agent/events.go [hunk 18]
```diff
@@ -141,15 +140,78 @@ func eventsProcess(event api.Event) {
...
+			return
+		}
+
+		file, err := os.Open("/proc/self/mountinfo")
+		if err != nil {
+			logger.Infof("Error opening /proc/self/mountinfo: %v\n", err)
```
Comment (tomponline) [inline lxd-agent/events.go:165]: Shall we make this a logger.Warnf?

File: lxd-agent/events.go [hunk 19]
```diff
@@ -141,15 +140,78 @@ func eventsProcess(event api.Event) {
...
+		for scanner.Scan() {
+			fields := strings.Fields(scanner.Text())
+			if len(fields) >= 10 && fields[4] == targetPath {
+				mountPoints = append(mountPoints, fields[4])
+			}
+		}
+
+		err = scanner.Err()
+		if err != nil {
+			logger.Infof("Error reading /proc/self/mountinfo: %v\n", err)
```
Comment (tomponline) [inline lxd-agent/events.go:182]: New \n in log messages please

File: lxd-agent/events.go (hunk 18 above)
Comment (tomponline) [inline lxd-agent/events.go:165]: New \n in log messages please

File: lxd-agent/events.go [hunk 20]
```diff
@@ -141,15 +140,78 @@ func eventsProcess(event api.Event) {
...
+			return
+		}
+
+		if len(mountPoints) == 0 {
+			logger.Infof("No mount points found for %s", targetPath)
+			return
+		}
+
+		if len(mountPoints) == 1 {
```
Comment (tomponline) [inline lxd-agent/events.go:191]: We shouldn't need this special case, if there is only a single entry the general loop below should still be able to work to deal with it.

File: lxd-agent/events.go [hunk 21]
```diff
@@ -141,15 +140,78 @@ func eventsProcess(event api.Event) {
...
+			return
+		}
+
+		defer file.Close()
+
+		var mountPoints []string
+		scanner := bufio.NewScanner(file)
+		for scanner.Scan() {
+			fields := strings.Fields(scanner.Text())
+			if len(fields) >= 10 && fields[4] == targetPath {
```
Comment (tomponline) [inline lxd-agent/events.go:175]: Should you match by prefix here so that mounts below the targetPath  are unmounted too?

File: lxd-agent/events.go (hunk 19 above)
Comment (tomponline) [reply lxd-agent/events.go:182]: Shall we make this a logger.Warnf?

File: lxd-agent/events.go [hunk 22]
```diff
@@ -141,15 +140,78 @@ func eventsProcess(event api.Event) {
...
+		}
+
+		err = scanner.Err()
+		if err != nil {
+			logger.Infof("Error reading /proc/self/mountinfo: %v\n", err)
+			return
+		}
+
+		if len(mountPoints) == 0 {
+			logger.Infof("No mount points found for %s", targetPath)
```
Comment (tomponline) [inline lxd-agent/events.go:187]: Shall we make this a logger.Warnf?

File: lxd-agent/events.go [hunk 23]
```diff
@@ -141,15 +140,78 @@ func eventsProcess(event api.Event) {
...
+		// Reverse the slice to unmount in reverse order.
+		// This is needed to unmount potential over-mounts first.
+		for i, j := 0, len(mountPoints)-1; i < j; i, j = i+1, j-1 {
+			mountPoints[i], mountPoints[j] = mountPoints[j], mountPoints[i]
+		}
+
+		for _, mountPoint := range mountPoints {
+			err = unix.Unmount(mountPoint, unix.MNT_DETACH)
+			if err != nil {
+				logger.Infof("Error unmounting %s: %v\n", mountPoint, err)
```
Comment (tomponline) [inline lxd-agent/events.go:210]: Shall we make this a logger.Warnf?

File: lxd/instance/drivers/driver_qemu.go [hunk 24]
```diff
@@ -2309,6 +2325,63 @@ func (d *qemu) deviceAttachBlockDevice(deviceName string, configCopy map[string]
...
 
+func (d *qemu) unmountPath(deviceName string, rawConfig deviceConfig.Device) error {
+	// First, cleanly unmount the path inside the VM.
+	msg := map[string]any{
+		"action": "removed",
+		"name":   deviceName,
+		"config": rawConfig,
+	}
+
+	err := d.devlxdEventSend("device", msg)
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:2336]: This actually seems like a problem now im looking at it because we are going to be sending a duplicated removed event, as its also sent once removal is finished (the correct time) in qemu.Update.

Perhaps we can send an `unmount` event?

File: lxd/instance/drivers/driver_qemu.go [hunk 25]
```diff
@@ -2309,6 +2325,63 @@ func (d *qemu) deviceAttachBlockDevice(deviceName string, configCopy map[string]
...
+	if err != nil {
+		return err
+	}
+
+	// Now that the unmount has occurred on the VM,
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:2341]: How do we know this? Perhaps a synchronous API event is what we need infact?

Or should the lxd-agent be responsible for cleaning up the mount and directories? 

Thoughts?

File: lxd/instance/drivers/driver_qemu.go [hunk 26]
```diff
@@ -2309,6 +2325,63 @@ func (d *qemu) deviceAttachBlockDevice(deviceName string, configCopy map[string]
 	return nil
 }
 
+func (d *qemu) unmountPath(deviceName string, rawConfig deviceConfig.Device) error {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:2328]: In general, does this still allow disk shares to be removed on a stopped VM, even though the clean up cannot occur?

File: lxd-agent/events.go (hunk 16 above)
Comment (gabrielmougard) [reply lxd-agent/events.go:150]: yes

File: lxd-agent/events.go (hunk 21 above)
Comment (gabrielmougard) [reply lxd-agent/events.go:175]: yes that's a good idea

File: lxd-agent/events.go (hunk 16 above)
Comment (gabrielmougard) [reply lxd-agent/events.go:150]: fixed

File: lxd-agent/events.go (hunk 18 above)
Comment (gabrielmougard) [reply lxd-agent/events.go:165]: fixed

Comment (gabrielmougard) [reply lxd-agent/events.go:165]: fixed

File: lxd-agent/events.go (hunk 21 above)
Comment (gabrielmougard) [reply lxd-agent/events.go:175]: fixed

File: lxd-agent/events.go (hunk 19 above)
Comment (gabrielmougard) [reply lxd-agent/events.go:182]: fixed

File: lxd-agent/events.go (hunk 22 above)
Comment (gabrielmougard) [reply lxd-agent/events.go:187]: fixed

File: lxd-agent/events.go (hunk 20 above)
Comment (gabrielmougard) [reply lxd-agent/events.go:191]: fixed

File: lxd-agent/events.go [hunk 27]
```diff
@@ -141,15 +140,78 @@ func eventsProcess(event api.Event) {
...
+
+		if len(mountPoints) == 0 {
+			logger.Infof("No mount points found for %s", targetPath)
+			return
+		}
+
+		if len(mountPoints) == 1 {
+			err = unix.Unmount(mountPoints[0], unix.MNT_DETACH)
+			if err != nil {
+				logger.Infof("Error unmounting %s: %v\n", mountPoints[0], err)
```
[the remark this answers is not in the document]
Comment (gabrielmougard) [reply lxd-agent/events.go:194]: fixed

File: lxd-agent/events.go (hunk 23 above)
Comment (gabrielmougard) [reply lxd-agent/events.go:210]: fixed

File: lxd/instance/drivers/driver_qemu.go (hunk 25 above)
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_qemu.go:2347]: @tomponline My thoughts on the second solution you suggests are: it seems easier and would require less changes, but I think there would be an issue in the end:

Since the `eventType` would be `unmount` as you asked above, I could pass a msg like:
```go
msg := map[string]any{
		"action": "removed",
		"name":   deviceName,
		"volatile": d.deviceVolatileGetFunc(deviceName)(),
		"config": rawConfig,
	}
``` 

and do the file remove logic inside the agent but in case where `last_state.created` is not present in the volatile key, I'll still need to do:

```go
err = d.deviceVolatileSetFunc(deviceName)(map[string]string{"last_state.created": ""})
if err != nil {
	return fmt.Errorf("Error updating volatile for the device: %w", err)
}
```

after the unmount occur successfully (but because this is not synchronous, when do I know when to do that ?)

File: doc/config_options.txt (hunk 4 above)
Comment (gabrielmougard) [reply doc/config_options.txt]: fixed

File: lxd/instance/drivers/driver_lxc.go (hunk 5 above)
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_lxc.go]: fixed

File: doc/api-extensions.md (hunk 12 above)
Comment (gabrielmougard) [reply doc/api-extensions.md]: fixed

File: lxd/instance/drivers/driver_lxc.go (hunk 13 above)
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_lxc.go:1738]: fixed

File: lxd/instance/drivers/driver_lxc.go [hunk 28]
```diff
@@ -1690,6 +1690,14 @@ func (d *lxc) deviceDetachNIC(configCopy map[string]string, netIF []deviceConfig
 // deviceHandleMounts live attaches or detaches mounts on a container.
 // If the mount DevPath is empty the mount action is treated as unmount.
 func (d *lxc) deviceHandleMounts(mounts []deviceConfig.MountEntryItem) error {
+	// Connect to files API.
```
[the remark this answers is not in the document]
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_lxc.go:1696]: fixed

File: lxd/instance/drivers/driver_lxc.go (hunk 14 above)
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_lxc.go]: fixed

File: lxd/instance/drivers/driver_lxc.go (hunk 15 above)
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_lxc.go]: fixed

File: lxd/instance/drivers/driver_qemu.go (hunk 25 above)
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_qemu.go:2347]: @tomponline other idea, specifically for the unmount event we could make the API synchronous and leave it async for the other actions:

```go
func eventsPost(d *Daemon, r *http.Request) response.Response {
	...

	// Handle device related actions locally.
	eventsProcess(event) // instead of `go eventsProcess(event)` 
	return response.SyncResponse(true, nil)
}

func eventsProcess(event api.Event) {
	...
	if e.Action == "unmount" {
	   ... // synchronous code
	   
	   return
	}
	
	// for other actions, async behaviour is used
	go func() {
	    if e.Action == "added" {
	       ...
	    }
	    
	    return
	}()
```

File: lxd/instance/drivers/driver_qemu.go (hunk 26 above)
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_qemu.go:2334]: I don't think so. I'll have to check but since I can't have access to the agent when the VM is stopped, I won't be able to make use of the `devlxdEventSend` function..

**Comment by gabrielmougard** [comment]
@tomponline a couple of thoughts on your questions.

**Comment by gabrielmougard** [comment]
@tomponline can you review this ? Here is the LXD-CI PR https://github.com/canonical/lxd-ci/pull/83 for the VM tests.

File: lxd/instance/drivers/driver_qemu.go (hunk 25 above)
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_qemu.go]: fixed

File: lxd/instance/drivers/driver_qemu.go (hunk 24 above)
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_qemu.go]: fixed

File: lxd/instance/drivers/driver_qemu.go (hunk 26 above)
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_qemu.go:2321]: fixed

**Comment by gabrielmougard** [comment]
@tomponline @roosterfish can you review this please?

File: lxd-agent/devices.go [hunk 29]
```diff
@@ -0,0 +1,98 @@
...
+		}
+	}
+
+	// Now that the unmount has occurred,
+	// check if we need to remove the target path.
+	if device.Volatile != nil {
+		path, ok := device.Volatile["last_state.created"]
+		if ok {
+			if targetPath == path {
+				err := os.Remove(path)
```
[the remark this answers is not in the document]
Comment (roosterfish) [reply lxd-agent/devices.go:89]: @gabrielmougard you could try using the `targetPath` here instead as this one gets checked for being absolute a few lines above.

I guess we could argue if this is a security risk but I am wondering if it will satisfy CodeQL.

[the remark this answers is not in the document]
Comment (gabrielmougard) [reply lxd-agent/devices.go]: fixed

**Comment by tomponline** [comment]
Please can you rebase

**Comment by gabrielmougard** [comment]
@tomponline updated

**Review by tomponline — CHANGES_REQUESTED** [review]
Following up from our meeting, we need to:

- Check for performance reductions with this approach when starting instances.
- Consider what happens when part of the target path exists but part of it doesn't. What gets deleted?

**Comment by tomponline** [comment]
@gabrielmougard any news on this one?

**Comment by tomponline** [comment]
needs a rebase too please

**Comment by gabrielmougard** [comment]
@tomponline updated

**Comment by tomponline** [comment]
@gabrielmougard in your tests were you stopping with `--force` btw?

**Comment by gabrielmougard** [comment]
> @gabrielmougard in your tests were you stopping with `--force` btw?

I don't think so... Which tests are you talking about?

**Comment by tomponline** [comment]
> aaah, the benchmark.. No, I'm not using `--force`

are you able to try that so we dont get the guest OS shutdown time in the results?

**Comment by gabrielmougard** [comment]
@tomponline It should be updated now

File: shared/util.go [hunk 30]
```diff
@@ -1521,3 +1522,87 @@ func IsMicroOVNUsed() bool {
...
+
+// EncodeRemoteAbsPathWithNonExistingDir splits a path into existing and non-existing parts and encodes it.
+func EncodeRemoteAbsPathWithNonExistingDir(c *sftp.Client, absPath string) (string, error) {
+	if !filepath.IsAbs(absPath) {
+		return "", fmt.Errorf("Path is not absolute")
+	}
+
+	// First resolve the path (handles /./, /../, symlinks, etc.)
+	// A user might provide a path with /./ or /../ in it, so we need to resolve it.
+	resolvedExistingPath, err := c.RealPath(absPath)
```
Comment (tomponline) [inline shared/util.go:1534]: @gabrielmougard please can you find out if this resolves symlinks or not?

Comment (gabrielmougard) [reply shared/util.go:1537]: Not it does not. It just canonicalizes the path (handles `/./` and `/../` to return a proper path) . Weirdly, it does not even an error in case the target path does not exist (/a/b/c wil return /a/b/c even if it does not exist in the container fs). I went on using c.Lstat that seems to be more what we want in this situation.

**Comment by tomponline** [comment]
Closing as will re-evaluate how to progress with the linked issues.
